#create bedpe with simple annotation
#https://github.com/PapenfussLab/gridss/issues/561

# if (!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")
#BiocManager::install("StructuralVariantAnnotation")
#install.packages("stringr")
library(VariantAnnotation)
library(StructuralVariantAnnotation)
library(stringr)

#' Simple SV type classifier
simpleEventType <- function(gr) {
  pgr = partner(gr)
  return(ifelse(seqnames(gr) != seqnames(pgr), "Trans", # inter-chromosomosal
    ifelse(strand(gr) == strand(pgr), "Inv",
      ifelse(gr$insLen >= abs(gr$svLen) * 0.7, "Ins", # TODO: improve classification of complex events
        ifelse(xor(start(gr) < start(pgr), strand(gr) == "-"), "Del",
          "Dup")))))
}

# using the example in the GRIDSS /example directory
args <- commandArgs(trailingOnly = TRUE)

vcf_file <- args[1]
out_dir <- args[2]
genome <- args[3]

vcf <- readVcf(vcf_file, genome)

info(header(vcf)) = unique(as(rbind(as.data.frame(info(header(vcf))), data.frame(
	row.names=c("SIMPLE_TYPE"),
	Number=c("1"),
	Type=c("String"),
	Description=c("Simple event type annotation based purely on breakend position and orientation."))), "DataFrame"))

#svLen is available in breakpointRanges
gr <- breakpointRanges(vcf, nominalPosition=TRUE)
svtype <- simpleEventType(gr)
# info(vcf)$SIMPLE_TYPE <- NA_character_
# info(vcf[gr$sourceId])$SIMPLE_TYPE <- svtype
# info(vcf[gr$sourceId])$SVLEN <- gr$svLen
#writeVcf(vcf, "high_confidence_somatic.sv.annotated.vcf") # generated by example/gridss.sh


bedpe <- data.frame(
    chrom1=seqnames(gr),
    start1=start(gr),
    end1=end(gr),
    chrom2=seqnames(partner(gr)),
    start2=start(partner(gr)),
    end2=end(partner(gr)),
    name=names(gr),
    score=gr$QUAL,
    strand1=strand(gr),
    strand2=strand(partner(gr)),
    CN_overlap_type=svtype
    )
bedpe$orient_1 <- ifelse(bedpe$strand1=="+","rev","fwd")
bedpe$orient_2 <- ifelse(bedpe$strand2=="+","rev","fwd")
bedpe$Sample <- tail(samples(header(vcf)),1) #last element from the header

# Just the lower of the two breakends so we don't output everything twice
bedpe <- bedpe[str_detect(bedpe$name, "gridss.+o"),]

#add SPAN info
bedpe$SPAN <- bedpe$start2-bedpe$start1
bedpe[which(bedpe$SPAN < 0),"SPAN"] <- -1

write.table(bedpe, paste0(out_dir,gsub(".vcf",".bedpe",basename(vcf_file))), quote=FALSE, sep='\t', row.names=FALSE, col.names=TRUE)


